### PR TITLE
Issue/7540 followed blog notifications

### PR DESCRIPTION
**Pull request by theck13**
## ~~Do Not Merge~~
~~This is dependent upon https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/774.  Once that is merged, the FluxC commit hash will need to be updated and these changes will be ready to merge following the code review.~~

### Fix
This is only ***Part 1 of 2*** for the [Notification for Followed Blogs issue](https://github.com/wordpress-mobile/WordPress-Android/issues/7540).  I divided the issue into two parts so the pull requests aren't too big.

Add followed blogs to Notification Settings allowing management of notifications for new post and comments.  See the screenshots below for illustration.

notifications_settings_list

notifications_settings_dialog_collapsed

notifications_settings_dialog_expanded

The `RadioGroup` in `followed_sites_dialog.xml` uses the `android:layoutDirection="rtl"` attribute in order to display the radio buttons to the right of the text as shown in the screenshots above.  Unfortunately, the `layoutDirection` attribute is not available on API 16 and below.  Consequently, the layouts will be different on API 16 compared to later versions.  See the screenshots below for illustration.

notification_settings_followed_dialog_api

### Test
1. Go to ***Me*** tab.
2. Tap ***Notification Settings*** item.
3. Notice ***Followed Sites*** section.
4. Tap site under ***Followed Sites*** section.
5. Edit notifications settings in dialog.
6. Tap ***OK*** button.
7. Tap site from Step 5.
8. Notice notifications settings are saved.
9. Edit notifications settings in dialog.
10. Tap ***Cancel*** button.
11. Tap site from Step 5.
12. Notice notifications settings are not saved.

### CODE REVIEW

**Comment by nbradbury** [comment]
@theck13 The first time I ran this build the list of followed sites in settings was empty. I went to the reader and noticed it showed all my followed sites, but when I returned to settings they still weren't showing (but they did eventually show up). 

This made me wonder if we're storing followed sites in two places - once in the ReaderBlogTable and again in FluxC - and I **think** this is the case. If so, we'll want to drop the old method of fetching & persisting followed sites and convert to the FluxC version (but not as part of this PR, especially since @mzorz is currently updating the app's reader services).

screenshot_1523099619

**Comment by nbradbury** [comment]
When I view the full list of "All My Followed Sites," they appear in random order. I expected them to be sorted by title.

**Comment by nbradbury** [comment]
If I turn off notifications, I'm told all notifications "from this app" will be disabled.  But I continue to get email notifications, and I can't turn them off until I re-enable notification settings. I think a user would get to this point and not know what to do (but at least the email notification includes an Unsubscribe link).

screenshot_1523123819

**Comment by nbradbury** [comment]
Just in case this isn't mentioned during the design review, I noticed some sites don't show a title but these same sites **do** show a title in the reader. Also, note how the reader shows "friendlier" URLs rather than the entire site address.

screenshot_1523124283

screenshot_1523124357

**Comment by nbradbury** [comment]
The snackbar when you follow a site in the reader is a nice touch, but I found the wording awkward. Also, should the same snackbar appear when you follow from elsewhere (reader detail & reader site preview)?

screenshot_1523125373

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java [hunk 1]
```diff
@@ -762,6 +763,26 @@ public void onEventMainThread(ReaderEvents.SearchPostsEnded event) {
...
+    private void showPostNotificationsSnackbar(ReaderPost post) {
+        if (getView() != null) {
+            Snackbar.make(getView(), Html.fromHtml(getString(R.string.reader_followed_blog_notifications,
+                        post.getBlogName())),
+                        Snackbar.LENGTH_LONG)
+                    .setAction(getString(R.string.reader_followed_blog_notifications_action),
+                        new View.OnClickListener() {
+                            @Override public void onClick(View view) {
+                                // TODO: Replace message with API call.
+                                Toast.makeText(getActivity(), "Enabled", Toast.LENGTH_SHORT).show();
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java]: Is the API call going to be added in this PR?

File: WordPress/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -1340,6 +1351,9 @@
     <string name="reader_btn_follow">Follow</string>
     <string name="reader_btn_unfollow">Following</string>
 
+    <string name="reader_followed_blog_notifications">Following &lt;b&gt;%1$s&lt;/b&gt;&lt;br&gt;Enable post notifications?</string>
```
Comment (nbradbury) [inline WordPress/src/main/res/values/strings.xml]: That looks tricky to translate! Might be simpler to use `Following %s\nEnable post notifications?` and have the app insert the HTML bolding in the `%s` replacement.

**Comment by nbradbury** [comment]
I should mention that I'm doing code reviews over the weekend to make up for time I'll be away from work later this week, so please don't feel the need to pull yourself away from your weekend just because I'm commenting here.

> Yeah, I expected the same as well and I was going to sort them accordingly, but I noticed Your Sites are not sorted alphabetically by title either. 

I hadn't noticed that, but now that I do I think sorting Your Sites should be filed as an issue and fixed, and likewise we should sort the followed sites.

> Email notifications are not "from this app" per se.

Oh, I agree, and I should've been more clear. To me the problem is that if you disable app notifications, you can't modify your followed sites email notifications even though they're unrelated to app notifications.

> ...a Snackbar shouldn't be more than a few lines of text and the blog title is a variable length not to mention translation complications as well. Thus, we need succinct wording, which can seem a little awkward.

Agreed about being succint, I just found it hard to read that snackbar and figure out what it meant - it all runs together.

> That's an unfortunate side effect of the dialog interface...we have to hit the API after the dialog is closed, which can take some time on slow connections

Do we not have the changes cached locally so we can reflect them when the dialog is shown?

> I added methods similar to the Reader behavior you mentioned for the site title and "friendlier" URLs

I think that's a **big** improvement, but in some cases it could use tweaking. Screenshot below showing the same subscription in the reader and notif settings. 

screenshot_1523142389

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationSettingsFollowedDialog.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    private String mRadioButtonSelected;
+    private Switch mSwitchEmailComments;
+    private Switch mSwitchEmailPosts;
+    private Switch mSwitchNotificationPosts;
+    private boolean mConfirmed;
+
+    @Override
+    public Dialog onCreateDialog(Bundle savedInstanceState) {
+        LayoutInflater inflater = getActivity().getLayoutInflater();
+        //noinspection InflateParams
```
[the remark this answers is not in the document]
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationSettingsFollowedDialog.java]: I remove the comment in 8ac4bd0.

**Comment by nbradbury** [comment]
> When a followed site is tapped in the list, the settings are retrieved from the SubscriptionModel, which is updated from the API callback. How do you think we should cache those changes?

Can we update the SubscriptionModel when the OK button is tapped, and then revert the change on failure?

**Comment by theck13** [comment]
> Can we update the SubscriptionModel when the OK button is tapped, and then revert the change on failure?

We'll have to keep a log of which subscriptions were changed and what the values were originally in order to revert.  That could be tricky with slow connections, multiple subscription changes, changing the same subscription multiple times, leaving the Notification Settings screen, etc.

**Comment by theck13** [comment]
The changes in FluxC have been merged, the dependency has been updated, and merge conflicts have been resolved.  This pull request is no longer in "Do Not Merge" status.

**Comment by SylvesterWilmott** [comment]
> If I turn off notifications, I'm told all notifications "from this app" will be disabled. But I continue to get email notifications, and I can't turn them off until I re-enable notification settings. I think a user would get to this point and not know what to do (but at least the email notification includes an Unsubscribe link).

This is an issue for notification settings in general. The fix would be to either kill emails when notifications are turned off (not too desirable) or keep email options available after the master app notifications has been disabled (preferred). I think both are relatively easy fixes but perhaps not for this particular task, @theck13 if you want to try to include in the next PR I can produce a design?


> When I view the full list of "All My Followed Sites," they appear in random order. I expected them to be sorted by title.

Sorted by title would be the ideal situation. @theck13 maybe we can open a new issue and include the previous one as well?


> I added methods similar to the Reader behavior you mentioned for the site title and "friendlier" URL's in ca5ddb8. However, there are still sites returned from the API without any title. In those cases, I'm using the URL.

This works fine.


> That's an unfortunate side effect of the dialog interface. We don't want to save the settings as they are toggled since the user can cancel the dialog and they would expect nothing to have changed even if they toggled something. Also, we don't necessarily want to have a blocking interface like a progress dialog after the user taps the OK button and confirms the changes. Due to that, we have to hit the API after the dialog is closed, which can take some time on slow connections as you mentioned.

I am not really expecting users to rapidly open/change/accept this dialog. But I will track this to be sure.


> Ahh, I see. Would something like "Enable site notifications for <SITE_NAME>?" be better?

We just talked about this but for the sake of transparency. The intention of the current design was to have two text views/labels which would truncate to 1 line by themselves. By doing this we avoid weird truncation issues or long site titles forcing important text beyond the 2 line system limit. I do like your suggestion though, by adding the site title to the end and only using 1 label which is also very succinct, we can probably dodge both bullets. I'll do some testing with different translations for this before we make a decision.

Overall the UI looks great. Excited to see this one coming together.

**Comment by theck13** [comment]
> This is an issue for notification settings in general. The fix would be to either kill emails when notifications are turned off (not too desirable) or keep email options available after the master app notifications has been disabled (preferred). I think both are relatively easy fixes but perhaps not for this particular task, @theck13 if you want to try to include in the next PR I can produce a design?

I agree; handling all email notification settings is outside of the scope of this task.  I say the three of us (i.e. @nbradbury, @SylvesterWilmott, and @theck13) chat about the options synchronously and create a new issue accordingly.

> Sorted by title would be the ideal situation. @theck13 maybe we can open a new issue and include the previous one as well?

I opened a [new issue](https://github.com/wordpress-mobile/WordPress-Android/issues/7660) to address them together outside of this pull request.

**Comment by nbradbury** [comment]
@theck13 Sounds like everything has either been addressed or moved to another issue, so if you're ready for me to say :shipit: just give me the word.

**Comment by theck13** [comment]
I think you're right, @nbradbury.  I'm ready when you are!

**Comment by nbradbury** [comment]
:shipit:

**Pull request merged**